recirq/optimize/mpg.py

```diff
@@ -0,0 +1,390 @@
+    )
+    shifted_xs = [(x - xopt) for x in xs]
+    model = model.fit(shifted_xs, ys)
+    return model
+
+
+def clip_by_norm(t, clip_norm):
+    """Clips tensor values to a maximum L2-norm.
+    returns a tensor of the same type and shape as `t` with its values set to:
+    `t * clip_norm / l2norm(t)`
+
+    Args:
+        t: The tensor values at which we want to clip the norm.
+        clip_norm: The scalar to give a maximal upper bound for the L2-norm of the tensor.
+
+    Returns:
+        the tensor with the same shape of the input, but with norm clipped.
+    """
+    l2_norm = np.linalg.norm(t, 2)
+    return t * clip_norm / np.maximum(l2_norm, clip_norm)
+
+
+class ConstantSchedule(object):
+    def __init__(self, value):
+        """The constant schedule for some hyperparameter (e.g. learning_rate)
+        
+        Value remains constant over time.
+
+        Args: 
+            value: Constant value of the schedule
+        
+        Returns: 
+            a class of the schedule
+        """
+        self._v = value
+
+    def value(self, t):
+        """Return the value of the schedule at time step t 
+
+        Args: 
+            t: the time step for the schedule 
+
+        Returns: 
+            the schedule value
+        """
+        return self._v
+
+
+class ExponentialSchedule(object):
+    def __init__(self, learning_rate, decay_steps, decay_rate, staircase=False):
+        """The Exponential schedule for some hyperparameter (e.g. learning_rate)
+        
+        Exponential decay for the `learning rate`. For each `decay_steps`, the learning 
+        rate is scheduled to decay at the `decay_rate`. The `staircase` controls whether 
+        to decay smoothly or discontinuously. After this many timesteps pass, the final
+        learning rate is returned.
+
+        Args: 
+            learning rate: the initial learning rate 
+            decay_steps: the learning rate is scheduled to decay every such number of steps 
+            decay_rate: the learning rate is scheduled to decay at the such rate
+            staircase: if True, the learning rate keeps the same before every decay steps; 
+                       otherwise, the learning rate decays smoothly according 
+                       to exponential interpolation.
+        
+        Returns: 
+            a class of the schedule
+        """
+        self.learning_rate = learning_rate
+        self.decay_steps = decay_steps
+        self.decay_rate = decay_rate
+        self.staircase = staircase
+
+    def value(self, t):
+        """Return the value of the schedule at time step t 
+
+        Args: 
+            t: the time step for the schedule 
+
+        Returns: 
+            the schedule value
+        """
+        m = t / self.decay_steps
+        if self.staircase == True:
+            m = np.floor(m)
+
+        return self.learning_rate * self.decay_rate ** m
+
+
+class Adam:
```

Comment:
why is this a class? just make it a function

Reply:
I meant `Adam`. `ExponentialSchedule` makes sense as a class since it closes over the hyperparameters (learning rate et. al.) and follows the (implicit) Schedule interface

Reply:
agreed!